lib/active_model/serializer/adapter/json_api/deserialization.rb

```diff
@@ -0,0 +1,207 @@
+module ActiveModel
+  class Serializer
+    module Adapter
+      class JsonApi
+        module Deserialization
+          module_function
+
+          # Transform a JSON API document, containing a single data object, from a hash
+          # into a hash that is ready for ActiveRecord::Base.new() and such.
+          # Raises ArgumentError 'Invalid payload' if the payload is not properly formatted.
+          #
+          # @param [Hash|ActionController::Parameters] document
+          # @param [Hash] options
+          #   fields: Array of symbols and a hash. Specify whitelisted fields, optionally
+          #     specifying the attribute name on the model.
+          # @return [Hash]
+          #
+          # @example
+          #   document = {
+          #     data: {
+          #       id: 1,
+          #       type: 'post',
+          #       attributes: {
+          #         title: 'Title 1',
+          #         date: '2015-12-20'
+          #       },
+          #       associations: {
+          #         author: {
+          #           data: {
+          #             type: 'user',
+          #             id: 2
+          #           }
+          #         },
+          #         second_author: {
+          #           data: nil
+          #         },
+          #         comments: {
+          #           data: [{
+          #             type: 'comment',
+          #             id: 3
+          #           },{
+          #             type: 'comment',
+          #             id: 4
+          #           }]
+          #         }
+          #       }
+          #     }
+          #   }
+          #
+          #   parse(document) #=>
+          #     # {
+          #     #   title: 'Title 1',
+          #     #   date: '2015-12-20',
+          #     #   author_id: 2,
+          #     #   second_author_id: nil
+          #     #   comment_ids: [3, 4]
+          #     # }
+          #
+          #   parse(document, fields: [:title, :author, date: :published_at]) #=>
+          #     # {
+          #     #   title: 'Title 1',
+          #     #   published_at: '2015-12-20',
+          #     #   author_id: 2,
+          #     # }
+          #
+          # rubocop:disable Metrics/CyclomaticComplexity
+          def parse!(document, options = {})
+            fields = parse_fields_option(options[:fields])
+
+            document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)
+
+            fail ArgumentError, 'Invalid payload' unless payload_valid?(document)
```

Comment:
could I intervene? Why not to use cons of both approachs?   

```ruby
DEFAULT_FALLBACK = ->(error) { fail ArgumentError, 'Invalid payload' }

def parse!(document, options = {}, &fallback)
  fallback ||= DEFAULT_FALLBACK
  
  fields = parse_fields_option(options[:fields])

  document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)

  fallback.call(document) unless payload_valid?(document)

  # ...
```

1) It will rise error as soon as possible wich is much more explicit way then simple return empty hash
2) It allows user to provide custom fallback which can be {}